### ISSUE TITLE
Final Racing Cyberspace 2001

### ISSUE THREAD

**Issue by AITUS95** [narzoul/DDrawCompat#271]
When I try to start a race, I get this messed-up graphics that won't let me do anything. I'm forced to exit the game.

![PIC](https://github.com/narzoul/DDrawCompat/assets/48457684/f3e78e31-b322-4482-a70b-e21bcb338874)

[DDrawCompat-racing.log](https://github.com/narzoul/DDrawCompat/files/13709320/DDrawCompat-racing.log)
[DDrawCompat-halh.log](https://github.com/narzoul/DDrawCompat/files/13709321/DDrawCompat-halh.log)

**Comment by narzoul** [comment]
Yeah, I can't get this game working either. Instead of that black screen with some HUD elements, I get a crash when loading a race. I've tried all combinations of the various patches (official patch, XP patch), CD copied to disk with registry update, etc., nothing works. It doesn't even run in a VM.

Does the game work for you without DDrawCompat? Have you tried rmh.exe also? That one is not hardware accelerated so it should be simpler to run, in theory. But I get the exact same crash with it as with halh.exe anyway.

I've no idea how to fix it, but from looking at the debug logs with DDrawCompat, it's not a DirectDraw/Direct3D issue.
I gave up trying after wasting several hours on it, I don't have time to reverse engineer the cause of the loading issue. The only curious thing I found is that the game creates a bunch of 0 size .x files upon launch, but maybe that happens for people who can run the game too, I couldn't find any info about it.

**Comment by AITUS95** [comment]
The game runs fine without ddrawcompat, but suddenly crashes during the race. I can only get this version to work with dgvoodoo2. This game is known to be a headache on modern systems. It turns out that if I install the game using the detailed redump version with the setup fix, it works smoothly with ddrawcompat. However, with the versions from myabandonware or old-game.ru, I encounter the mentioned errors. Here's the link where you can find the detailed version that works with ddrawcompat: [link](https://drive.google.com/file/d/1l-DDng6qbzf6ku5mIOR-9E7uYsvYPf9X/view?usp=drivesdk). I have a hunch that the ripped version might have been slightly altered to work on modern systems through another patch. Unlike the detailed version, it seems the ripped version has all tracks unlocked.

**Comment by BEENNath58** [comment]
I tried the game on different hardware, opposing CPUs, WinXP, 7, 11. Crashes exactly the same as soon as I launch an exe.

The game seems to be affected by something not related to DDraw, it crashes at a different stage

**Comment by BEENNath58** [comment]
I finall got the game to run. The game files had to be placed at C:/racing/racing

I had to put RacingCD and Racingdir keys at HKLM/SOFTWARE/FRONT FAREAST/RACING/
, values being C:/racing and C:/racing/racing respectively
(Yes native SOFTWARE, not Wow6432Node)

On my Windows 7, the graphics are glitched at the Hall and rml.exe s which is the low resolution executable, using DDrawCompat. Natively only hall exe is glitches

**Comment by BEENNath58** [comment]
The game devs were unfortunately, very near sighted. The game has the same issues on Windows 98 as it has on Windows 11.

The game needs be installed in a drive in the following order:
`*driveletter*:\racing\racing\`

Basically copy the CD directly and paste it in the root folder of a drive. So your exe's are located in the 2nd order racing folder

You have to make registry keys at HKCU\Software\Front Fareast\Racing\:
Racingcd=*driveletter*:\racing\
Racingdir=*driveletter*:\racing\racing

**Comment by BEENNath58** [comment]
I made a comparision of what every version of the game does. You can have a look.
https://sourceforge.net/p/dxwnd/discussion/general/thread/123d344b28/?page=1&limit=25#87b3

With the RIP the game works fine with DDC

**Comment by narzoul** [comment]
For me, the game works well without any patches by just copying the CD contents to wherever, and pointing the Racingcd registry key at it, and the Racingdir at the Racing subdirectory. Run the game with Racing\halh.exe.
When using the original CD drive, the game throws an assertion failure at startup because it tries to open CARCMA.VUE with read/only access, which won't work on a read-only drive.

FpsLimiter=msgloop(30) seems to work well in the race and also appears to fix the camera panning bug.

**Comment by BEENNath58** [comment]
Yes a 30FPS lock works as well an independent strategy.

With the general installation + official patch + some patched files from here: https://limewire.com/d/m3sYl#LMT9xQaFyv

The game seems to be working in its full form.

There seems to be some issues with low resolution modes, but I will check tomorrow and tell.

**Comment by BEENNath58** [comment]
I tried the game with DDC 0.6.0 on Win7. Unfortunately it brings some problems

Halh.exe: Screen flashes. Doesn't happen natively

https://github.com/user-attachments/assets/a570b0f5-5650-4c96-8218-f7d2fedf4712

Hall.ex: black screen. Same natively

https://github.com/user-attachments/assets/a6326682-50c2-4630-a0ad-7f5459e682fb

Rmh.exe: White shadows. Black natively

![Image](https://github.com/user-attachments/assets/be91676e-ac96-47db-87de-93a9b59fe949)

Rml.exe: White screen. Correctly rendered natively

https://github.com/user-attachments/assets/5be9b505-2eda-48d2-9102-f5ca702f9866

**Comment by narzoul** [comment]
> Halh.exe: Screen flashes. Doesn't happen natively

For me, only the colons in the times at the top flicker. You might be using a different version/patch. Anyway, it flickers natively too, it just flickers too fast without an FPS limit so it might just look dimmer instead of flickery. Try turning off or adjusting the FpsLimiter and compare it that way. It flickers with dgVoodoo too (if you use its own FPS limiter) so I don't think it's a bug in both wrappers.

The game has other strange bugs, for example the lap counter just shows "/2" or "/3" at start (not even always the same number for the same track), then a little while into the race it jumps to 2/2 or 2/3 or 3/3 immediately, and the colons in the "best time" stop flickering. Then, if you finish a lap like that, the lap counter goes back to 1/2 or 1/3 and the game crashes.
The low res version has flickering in the position and lap counters instead. Sometimes the position rapidly flips between 5/6 and 6/6 even though you're obviously in the last place with no opponents around.
I'm pretty sure this game is just full of random bugs like this, unrelated to ddraw.

> Hall.ex: black screen. Same natively

The low resolution versions use 512x384 so make sure to add it to SupportedResolutions. Works fine for me, other than the already mentioned bugs.

> Rmh.exe: White shadows. Black natively

It's a difference between SoftwareDevice=rgb and ramp. Just use SoftwareDevice=app for this. On Windows 10/11, you'll need an old enough version of d3dim.dll to make it work properly due to the broken ramp device.

> Rml.exe: White screen. Correctly rendered natively

Natively it also has bugs for me, the ready screen is broken and the countdown at the start is not transparent, for example.
But it does render correctly with dgVoodoo and some older versions of DDrawCompat.
It's an old optimization I added that seems to be causing the problem. I don't fully understand it, but I found a fix for it:
[ddraw.zip](https://github.com/user-attachments/files/19727494/ddraw.zip) ([diff.txt](https://github.com/user-attachments/files/19727495/diff.txt) compared to v0.6.0)

**Comment by BEENNath58** [comment]
Rml.exe works (white screen)
Rmh.exe works (white shadow)
Halh.exe, removing the fpslimit doesn't flash

Hall.exe is still black

In the ini I have,
SoftwareDevice=app
SupportedResolutions=native, 512x384

Also if it matters, I already have 512x384 in the driver

**Comment by narzoul** [comment]
Are you using the "patched" version? How did you install the game exactly and from which source?
I tried installing the patch from myabandonware, but that just crashes all of the executables on the load screen. FTH doesn't help.
Additionally, RMH has flashing and graphical glitches in the menus, and RML just flashes between the last screen from the video and a black screen.

I'm sure the remaining problems are caused by heap corruptions and not by DDrawCompat, which doesn't even hook into any of the software renderers, so it can't really cause these problems. Now I'm starting to doubt if the white shadow is really caused by RGB and Ramp differences too - more likely it's a failed texture load from yet another bug in the game (might explain also why in dgVoodoo the car shadows are a mirror of the car texture instead of black). And possibly my optimization wasn't at fault either after all, which would explain why I couldn't understand how it causes such problems...

**Comment by BEENNath58** [comment]
> Are you using the "patched" version? How did you install the game exactly and from which source?

Depends on how you call it patched. First copied the game CD to C:\ root, on top of it the official patch (not XP patch), on top of it the external patch that allows the game to slow down and run correctly (FPS fix) on modern OS. 

https://community.pcgamingwiki.com/files/file/3367-final-racing-cyberspace-2001-modern-windows-fix-speed-fix/

Then I edited the registry to point RacingCD at C:\Racing\ and Racingdir at C:\Racing\Racing\

> FTH doesn't help.

Never helped. It is likely the registry since the game is location hardcoded when you copy the files or use the modern patch I linked

> Now I'm starting to doubt if the white shadow is really caused by RGB and Ramp differences too - more likely it's a failed texture load from yet another bug in the game (might explain also why in dgVoodoo the car shadows are a mirror of the car texture instead of black).

As I am writing, I tested native Windows 11. 

The shadow is correctly black in Rmh.exe with the old d3dim.dll
Rml.exe natively renders the correct scene instead of being white.

The only difference being Halh.exe which doesn't flicker here on top?

>I'm sure the remaining problems are caused by heap corruptions and not by DDrawCompat, which doesn't even hook into any of the software renderers, so it can't really cause these problems.

Hall.exe is still black. I am suspecting the exe for this one, since the CD exe still works

I didn't test original game + official patch. Original game works, official patch crashes it on Win11 (but not on Win7)

**Comment by narzoul** [comment]
> Depends on how you call it patched. First copied the game CD to C:\ root, on top of it the official patch (not XP patch), on top of it the external patch that allows the game to slow down and run correctly (FPS fix) on modern OS.
> 
> https://community.pcgamingwiki.com/files/file/3367-final-racing-cyberspace-2001-modern-windows-fix-speed-fix/
> 
> Then I edited the registry to point RacingCD at C:\Racing\ and Racingdir at C:\Racing\Racing\

You seem to be missing some steps here. If I copy the CD to C:\ then I'll only have a C:\Racing directory, but not C:\Racing\Racing. Same if I then use the patched setup to try to install the game from C:\setup.exe. It will actually complain about overwriting some files in C:\Racing with themselves, and there is only a Cancel button.

How did you get a C:\Racing\Racing directory? What puts it there?

**Comment by BEENNath58** [comment]
By copying the CD, I mean to right click and copy the CD drive so that you get a RACING folder in C

In RACING folder, another Racing folder contains the exes.

Cracyc found this:
https://sourceforge.net/p/dxwnd/discussion/general/thread/123d344b28/#3239/910d/3c32/7685/1b38/11d7

With the installer, the game puts some files extracted from somewhere which also allows portability. With this level of installation the game can be moved anywhere (withing a 50 char path length) and editing the reg.

However when you install the last "modern patch" (fps limiter and those stuff), they excluded the reading of the "some extracted files" so you get locked to C:\Racing\Racing\ and C:\Racing\ respectively

**Comment by BEENNath58** [comment]
> For me, the shadows are fixed
> A bisect still points to the same optimization,

Yes it already worked on your last suggestion. I just wrote in brackets for ease of remembering.

Only Halh.exe remember for now

**Comment by narzoul** [comment]
Well, it turns out the flickering is caused by a WM_ERASEBKGND message erasing the screen to black on the primary surface that is shared with GDI. But the behavior is the same even on a Windows XP VMware machine (apart from the fact that the 3d graphics is also broken there, but that's beside the point), so I don't think it's a bug in DDrawCompat. More likely a side effect of the XP patch, since the original game doesn't have it. Possibly it could be easy to patch by setting the main window class' background brush to null, I'll give it a go.

**Comment by narzoul** [comment]
Yeah, it's a simple patch in halh.exe:
`04 FF 15 2C 23 5B 00` -> `05 FF 15 2C 23 5B 00`

**Comment by BEENNath58** [comment]
The hex edit fixes the game with DDraw.

BUT, why does this problem not happen natively, in Windows 11?

https://github.com/user-attachments/assets/4c63a6d9-71b9-425a-af9d-a70bc727e22d

**Comment by narzoul** [comment]
> BUT, why does this problem not happen natively, in Windows 11?

I assume because DWM breaks GDI rendering. Maybe try Windows 7 with DWM on and off.

> Also any ideas as to what might have happened to hall.exe

But patch probably. I'm not really interested in looking into it considering it's broken even on XP, and it's caused by some unofficial mod/patch.

**Comment by BEENNath58** [comment]
> I assume because DWM breaks GDI rendering. Maybe try Windows 7 with DWM on and off.

How do I make DWM not turn off? Since it sees 16 bit ask from the game and turns off DWM.

Technically it could also run on 32bit colours, with DxWnd it runs on 16 and 32bit colours but crashes on 8 and 24. But how do I do it natively?

**Comment by narzoul** [comment]
Ah, I don't know actually. But since it turns off, is it flashing on Windows 7?

**Comment by narzoul** [comment]
If it's still flashing on Windows 7, maybe it's better to test Windows 8 next. There DWM is always on by default, but supposedly it can be turned off with some hacks, so it might be easier to compare on the same platform. But I'm not sure if this game is really worth that much hassle.

**Comment by BEENNath58** [comment]
It flashes on Win7. I don't have a Win8 setup to test now, I will test it when I get a test system but not now.

For now the work is done, the thread can be kept opened for later updates or the OPs doubts

**Comment by BEENNath58** [comment]
I installed Windows 8.1, but I can't get the game to work natively with DDrawcompat. It crashes on a black screen.

Using DxWnd Handle exceptions it works, in fullscreen as well.

I am yet to find how to disable DWM

**Comment by narzoul** [comment]
The patch for RML was added to v0.7.0. The rest of the issues are out of scope.
